File: examples/java/customer/src/main/java/dev/angzarr/examples/customer/CommandValidationException.java
package dev.angzarr.examples.customer;

import io.grpc.Status;

/**
 * Exception thrown when command validation fails.
 */
public class CommandValidationException extends Exception {
    private final Status.Code statusCode;

    public CommandValidationException(Status.Code statusCode, String message) {
        super(message);
        this.statusCode = statusCode;
    }

    public static CommandValidationException invalidArgument(String message) {
        return new CommandValidationException(Status.Code.INVALID_ARGUMENT, message);
    }

    public static CommandValidationException failedPrecondition(String message) {
        return new CommandValidationException(Status.Code.FAILED_PRECONDITION, message);
    }

    public Status.Code getStatusCode() {
        return statusCode;
    }
}

File: examples/kotlin/customer/src/main/kotlin/dev/angzarr/examples/customer/CommandValidationException.kt
package dev.angzarr.examples.customer

import io.grpc.Status

/**
 * Exception for command validation failures.
 * Bridges domain validation to gRPC status codes without coupling logic to gRPC.
 */
class CommandValidationException(
    message: String,
    val statusCode: Status.Code
) : Exception(message) {

    companion object {
        fun invalidArgument(message: String): CommandValidationException =
            CommandValidationException(message, Status.Code.INVALID_ARGUMENT)

        fun failedPrecondition(message: String): CommandValidationException =
            CommandValidationException(message, Status.Code.FAILED_PRECONDITION)
    }
}
